### ISSUE TITLE
Migration to `filepath.WalkDir` breaks symlinked Style directories

### ISSUE THREAD

**Issue by lykinsbd** [jdkato/txtlint#384]
# Issue Encountered

The changes implemented [here](https://github.com/errata-ai/vale/commit/8fd4450dfc2bec763ac9a2c8a897908eb0850a4f) by @jdkato in order to migrate off of `godirwalk` to `filepath.WalkDir` break Vale for any users who have symbolic links in place for their Style directories.

This is because of intentional design in the the underlying library ([filepath.WalkDir](https://pkg.go.dev/path/filepath#WalkDir)):
> WalkDir does not follow symbolic links.

Also see this comment by an author https://github.com/golang/go/issues/4759#issuecomment-66074349

## Use Case for Symlinks

The use case for having symbolic links to your Style directories is to allow multiple upstream Styles to be kept synced via Git in their respective file system locations. Then you symlink the different Git managed style directories into the Styles directory specified in `.vale.ini`.

# Examples and Details

Vale CLI doesn't provide any output to the user to indicate this failure is happening, it completes successfully but simply shows no errors/warnings/suggestions. Example shown below.

## Vale Version

```shell
$ vale --version
vale version 2.13.0
```

## Vale INI

```ini
StylesPath = /opt/venv/vale_styles/styles
MinAlertLevel = suggestion

[*.md]
BasedOnStyles = Microsoft, write-good, proselint, alex
write-good.E-Prime = NO
```

## Styles Directory Output

This is the directory with all 4 configured styles symlinked to the Git repositories where they are being pulled from upstream.

```shell
$ ll /opt/venv/vale_styles/styles
total 0
drwxr-xr-x  6 lykinsbd  admin   192B Sep 13 15:56 ./
drwxr-xr-x  8 lykinsbd  admin   256B Sep 15 20:44 ../
lrwxr-xr-x  1 lykinsbd  admin    22B Sep 13 15:55 Microsoft@ -> ../Microsoft/Microsoft
lrwxr-xr-x  1 lykinsbd  admin    12B Sep 13 15:55 alex@ -> ../alex/alex
lrwxr-xr-x  1 lykinsbd  admin    22B Sep 13 15:56 proselint@ -> ../proselint/proselint
lrwxr-xr-x  1 lykinsbd  admin    24B Sep 13 15:56 write-good@ -> ../write-good/write-good
```

## Example Failure

```shell
vale --config /opt/venv/vale_styles/.vale.ini README.md
✔ 0 errors, 0 warnings and 0 suggestions in 1 file.
```

## Prove symlinks are the problem

I made a copy of the Microsoft style into the styles directory (instead of symlinking the directory), and re-ran vale:

```shell
$ rm Microsoft
remove Microsoft? y
$ cp -r ../Microsoft/Microsoft .
$ ll 
total 0
drwxr-xr-x   6 lykinsbd  admin   192B Dec  2 10:15 ./
drwxr-xr-x   8 lykinsbd  admin   256B Sep 15 20:44 ../
drwxr-xr-x  45 lykinsbd  admin   1.4K Dec  2 10:15 Microsoft/
lrwxr-xr-x   1 lykinsbd  admin    12B Sep 13 15:55 alex@ -> ../alex/alex
lrwxr-xr-x   1 lykinsbd  admin    22B Sep 13 15:56 proselint@ -> ../proselint/proselint
lrwxr-xr-x   1 lykinsbd  admin    24B Sep 13 15:56 write-good@ -> ../write-good/write-good
$ vale --config /opt/venv/vale_styles/.vale.ini README.md
<snip many problems!>
✖ 11 errors, 5 warnings and 26 suggestions in 1 file.
```

# Proposed Solutions

I see two possible solutions, as outlined below.

1. Implement a check for if a directory contains symlinks. If the directory does contain symlinks, utilize a slower/less optimized walk method to gather the styles.
2. Document that symbolic links for Style directories are not supported in Vale.

I am happy to take a pass at a PR for option number 1 if it is desired.  But if the desire is simply to document that thou-shalt-not-use-symlinks in the documentation somewhere, I'm fine with that too.

Thank you!

-Brett

**Comment by ChrisChinchilla** [comment]
I can confirm the same issue, as I also make heavy use of symlinks, and believe I was the one who requested support for them in the first place 😅

**Comment by ChrisChinchilla** [comment]
Actually, now I am not so sure, I think it is still working

**Comment by jdkato** [comment]
Yes, this was an oversight on my part.

I'm open to option (1); it could be as simple as reverting the changes in `internal/core/util.go` and `internal/check/manager.go`, but leaving the others.

**Comment by izuzak** [comment]
👋 I was just setting up Vale for the first time and ran into this problem. Was pretty confused by things not working and then by accident found that they worked when there's no symlink. Not a major problem, but would be neat to make this work so just wanted to add my 👍 for this issue. :v:
